markdown/4_BayesHmm.md

```diff
@@ -0,0 +1,310 @@
+---
+title: Bayesian Hidden Markov Models
+permalink: /:collection/:name/
+---
+
+
+
+This tutorial illustrates training Bayesian Hidden Markov Models (HMM) using *Turing*. The main goals are learning the transition matrix, emission parameter, and hidden states.
+
+Let's load the libraries we'll need. We also set a random seed (for reproducibility) and the automatic differentiation backend to forward mode (more [here](https://github.com/TuringLang/Turing.jl/issues/511) on why this is useful). 
+
+````julia
+# Load libraries.
+using Turing, Plots, Random
+
+# Set a random seed and use the forward_diff AD mode.
+Random.seed!(1234);
+Turing.setadbackend(:forward_diff);
+````
+
+
+
+
+## Simple State Detection
+
+In this example, we'll use something where the states and emission parameters are very obvious. 
+
+````julia
+# Define the emission parameter.
+y = [ 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 2.0, 2.0, 2.0, 1.0, 1.0 ];
+N = length(y);  K = 3;
+
+# Plot the data we just made.
+plot(y, xlim = (0,15), ylim = (-1,5), size = (500, 250))
+````
+
+
+![](/tutorials/figures/4_BayesHmm_2_1.svg)
+
+
+We can see that we have three states, one for each height of the plot (1, 2, 3). This height is also our emission parameter, so state one produces a value of one, state two produces a value of two, and so on.
+
+Ultimately, we would like to understand three major parameters:
+
+1. The transition matrix. This is a matrix that assigns a probability of switching from one state to any other state, including the state that we are already in.
+2. The emission matrix, which describes a typical value emitted by some state. In the plot above, the emission parameter for state one is simply one.
+3. The state sequence is our understanding of what state we were actually in when we observed some data. This is very important in more sophisticated HMM models.
+
+With this in mind, let's set up our model.
+
+````julia
+# Turing model definition.
+@model BayesHmm(y) = begin
+    # State sequence.
+    s = tzeros(Int, N)
```

Comment:
`tzeros(Int, N)` creates a `TArray{Int}` of length `N`. The `TArray` is required here because we are using the `PG` sampler. 

There's some more info on that [here](http://turing.ml/docs/guide/#working-with-libtaskjl).

Reply:
Good to know -- I was wondering what that was for. Thanks for the pointer.